universe/ValueRefs.h

```diff
@@ -1695,34 +1633,53 @@ Operation<T>::Operation(OpType op_type, std::vector<std::unique_ptr<ValueRef<T>>
     m_op_type(op_type),
     m_operands(std::move(operands))
 {
-    DetermineIfConstantExpr();
+    InitConstInvariants();
     CacheConstValue();
 }
 
 template <typename T>
-void Operation<T>::DetermineIfConstantExpr()
+void Operation<T>::InitConstInvariants()
 {
     if (m_op_type == RANDOM_UNIFORM || m_op_type == RANDOM_PICK) {
         m_constant_expr = false;
+        this->m_root_candidate_invariant = false;
+        this->m_local_candidate_invariant = false;
+        this->m_target_invariant = false;
+        this->m_source_invariant = false;
         return;
     }
 
-    m_constant_expr = true; // may be overridden...
+    m_constant_expr = std::all_of(m_operands.begin(), m_operands.end(),
+        [](const auto& operand) { return operand && operand->ConstantExpr(); });
 
-    for (auto& operand : m_operands) {
-        if (operand && !operand->ConstantExpr()) {
-            m_constant_expr = false;
-            return;
-        }
-    }
+    this->m_root_candidate_invariant = std::all_of(m_operands.begin(), m_operands.end(),
+        [](const auto& operand) { return operand && operand->RootCandidateInvariant(); });
+    this->m_local_candidate_invariant = std::all_of(m_operands.begin(), m_operands.end(),
+        [](const auto& operand) { return operand && operand->LocalCandidateInvariant(); });
+    this->m_target_invariant = std::all_of(m_operands.begin(), m_operands.end(),
+        [](const auto& operand) { return operand && operand->TargetInvariant(); });
+    this->m_source_invariant = std::all_of(m_operands.begin(), m_operands.end(),
+        [](const auto& operand) { return operand && operand->SourceInvariant(); });
+
+    // determine if this is a simple incrment operation
+    if (m_op_type != PLUS && m_op_type != MINUS)
```

Comment:
While this seems to be correct it is strange to read.  The previous code assigned some values to variables based on conditions.  Here you are leaving the function early based on conditions, the previous implementation had at least return true or false, which made it easier to grasp.  I would suggest to change it either to:

* Set `m_simple_increment` to false before each return statement.
* Assign the combination of all conditions to `m_simple_increment` with positive notation.

Reply:
How about explicitly default initial setting `m_simple_increment = false;` above this block?

Reply:
That would be better than the current implementation, but still no good IMO.

Reply:
hopefully acceptable now?